external/loaders/loaders/mappers/_hybrid.py

```diff
@@ -313,52 +321,120 @@ def _add_fine_res_inputs(fine: xarray.Dataset, input_feature_url: Optional[str]
     return merged.astype(numpy.float32)
 
 
-@mapper_functions.register
-def open_fine_resolution(
-    fine_url: str = "",
-    input_feature_url: Optional[str] = None,
-    include_temperature_nudging: bool = False,
-) -> GeoMapper:
-    """
-    Open the fine-res mapper optionally using state from another run.
+def open_fine_resolution_additional_source(
+    fine_url: str,
+    input_feature_url: str,
+    additional_source_url: str,
+    include_temperature_nudging: bool,
+) -> xarray.Dataset:
 
-    Args:
-        fine_url: url where coarsened fine resolution data is stored
-        input_feature_url: url to fv3gfs_run to use for state inputs. If not provided,
-            the state of the fine-res data will be used as input.
-        include_temperature_nudging: whether to include fine-res nudging in Q1
+    fine = _compute_fine_res_portion(fine_url, include_temperature_nudging)
+    additional_source = open_zarr_maybe_consolidated(additional_source_url)
 
-    Returns:
-        a mapper
-    """
-    return XarrayMapper(
-        open_fine_resolution_dataset(
-            fine_url=fine_url,
-            input_feature_url=input_feature_url,
-            include_temperature_nudging=include_temperature_nudging,
-        )
+    fine["Q1"] = fine.Q1 + additional_source.Q1
+    fine["Q2"] = fine.Q2 + additional_source.Q2
+
+    return _add_fine_res_inputs(fine, input_feature_url)
+
+
+def _extend_lower(fine: xarray.Dataset, vertical_dim: str = "z") -> xarray.Dataset:
+    fine_sources = get_variables(fine, ["Q1", "Q2"])
+    fine_sources_new_bottom = fine_sources.isel({vertical_dim: -2})
+    fine_sources_without_bottom = fine_sources.isel({vertical_dim: slice(None, -2)})
+    fine_sources_extended_lower = xarray.concat(
+        [fine_sources_without_bottom, fine_sources_new_bottom], dim=vertical_dim
     )
+    return fine_sources_extended_lower.Q1, fine_sources_extended_lower.Q2
 
 
-@mapper_functions.register
-def open_precomputed_fine_resolution(
-    fine_url: str, input_feature_url: Optional[str] = None
-) -> GeoMapper:
-    """
-    Open the fine-res mapper from precomputed data, optionally using state
-        from another run.
+def open_fine_resolution_extend_lower(
+    fine_url: str, input_feature_url: str, include_temperature_nudging: bool
+) -> xarray.Dataset:
+
+    fine = _compute_fine_res_portion(fine_url, include_temperature_nudging)
+    fine["Q1"], fine["Q2"] = _extend_lower(fine)
 
+    return _add_fine_res_inputs(fine, input_feature_url)
+
+
+FINE_RESOLUTION_APPROACHES = [
+    "apparent_sources_only",
+    "precomputed_apparent_sources",
+    "apparent_sources_plus_nudging",
+    "apparent_sources_plus_dynamics_differences",
+    "apparent_sources_extend_lower",
+]
+
+
+@dataclass
+class FineResolutionOpenerConfig:
+    approach: str
```

Comment:
```suggestion
    approach: Literal[*FINE_RESOLUTION_APPROACHES] = "apparent_sources_only"
```
could simplify your default handling and error checking. I think we should be able to use `Literal` now that we're on py3.8